src/main/java/io/github/thebusybiscuit/cscorelib2/collections/CopyUtils.java

```diff
@@ -0,0 +1,166 @@
+package io.github.thebusybiscuit.cscorelib2.collections;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import java.util.function.Supplier;
+
+/**
+ * Utilities for deep-copying collections, maps and arrays
+ */
+public final class CopyUtils {
+    
+    private CopyUtils() {}
+
+    /**
+     * Perform a deep copy of all the elements from a given {@link Collection} to another.
+     * <p>
+     * If the source collection contains <code>null</code> elements, the cloning function should be
+     * handle null input. Additionally, the sink should also accept any values
+     * contained by the source collection, null or not.
+     *
+     * @param source          The source of the elements
+     * @param cloningFunction The function which clones the elements
+     * @param sink            The collection in which to copy to the cloned elements to
+     * @param <T>             The type of elements in the collections
+     * @param <C>             The type of the sink
+     *
+     */
+    public static <T, C extends Collection<T>> void deepCopy(@NonNull Collection<T> source, @NonNull Function<T, T> cloningFunction, @NonNull C sink) {
```

Comment:
Should also annotate the output as `@Nonnull`

Reply:
This is a void method